### PR TITLE
Implement @file-services/resolve

### PR DESCRIPTION
**Pull request by AviVahl**

### CODE REVIEW

File: packages/resolve/test/request-resolver.spec.ts [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+                    'entry.js': EMPTY,
+                    'browser.js': EMPTY
+                }
+            })
+            const resolveRequest = createRequestResolver({ fs })
+
+            expect(resolveRequest('/', './lodash')).to.be.resolvedTo('/lodash/browser.js')
+        })
+
+        it('uses "browser" even if "main" was not defined', () => {
```
Comment (tomrav) [inline packages/resolve/test/request-resolver.spec.ts]:
```suggestion
        it('uses "browser" if "main" was not defined', () => {
```

File: packages/resolve/test/request-resolver.spec.ts [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+                    'package.json': '{"browser": "file.js"}',
+                    'file.js': EMPTY
+                }
+            })
+            const resolveRequest = createRequestResolver({ fs })
+
+            expect(resolveRequest('/', './lodash')).to.be.resolvedTo('/lodash/file.js')
+        })
+
+        it('prefers "main" when resolution "target" is set to "node"', () => {
```
Comment (tomrav) [inline packages/resolve/test/request-resolver.spec.ts]: Should there be a complementing test for this checking `target: 'browser'`?

File: packages/resolve/src/request-resolver.ts [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+            while (lastPath !== currentPath) {
+                const isPackagesRoot = basename(currentPath) === packageRoot
+                const packagesPath = isPackagesRoot ? currentPath : resolve(currentPath, packageRoot)
+                const requestInPackages = join(packagesPath, request)
+                const resolved = resolveAsFile(requestInPackages) || resolveAsDirectory(requestInPackages)
+                if (resolved) {
+                    return resolved
+                }
+                lastPath = currentPath
+                currentPath = isPackagesRoot ? dirname(dirname(currentPath)) : dirname(currentPath)
```
Comment (tomrav) [inline packages/resolve/src/request-resolver.ts]: Please add a comment why the `dirname(dirname...))`

File: packages/resolve/src/types.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    request: string
+) => IResolutionOutput | null
+
+/**
+ * Required fs APIs for request resolution.
+ * Currently a subset of the sync base file system API.
+ */
+export interface IResolutionFileSystem {
+    path: IFileSystemPath
+    realpathSync: IFileSystemSync['realpathSync']
```
Comment (tomrav) [inline packages/resolve/src/types.ts]: This is not used. Please remove.

File: packages/resolve/src/request-resolver.ts (hunk 3 above)
Comment (AviVahl) [reply packages/resolve/src/request-resolver.ts]: added comment

File: packages/resolve/src/types.ts (hunk 4 above)
Comment (AviVahl) [reply packages/resolve/src/types.ts]: removed

File: packages/resolve/test/request-resolver.spec.ts (hunk 2 above)
Comment (AviVahl) [reply packages/resolve/test/request-resolver.spec.ts]: added test, even though it's the default behavior and tested by first test as well. new test checks what happens if the user explicitly provided flag.

File: packages/resolve/test/request-resolver.spec.ts [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+                with_ext: {
+                    'package.json': '{"main": "entry.js"}',
+                    'entry.js': EMPTY
+                },
+                without_ext: {
+                    'package.json': '{"main": "main_file"}',
+                    'main_file.js': EMPTY
+                },
+                to_file_in_folder: {
+                    'inner': { 'file.js': EMPTY },
```
[the remark this answers is not in the document]
Comment (AviVahl) [reply packages/resolve/test/request-resolver.spec.ts]: swapped

File: packages/resolve/test/request-resolver.spec.ts [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+            const fs = createMemoryFs({
+                'src': {
+                    'file.js': EMPTY
+                },
+                'another.js': EMPTY
+            })
+            const resolveRequest = createRequestResolver({ fs })
+
+            expect(resolveRequest('/', './src/file')).to.be.resolvedTo('/src/file.js')
+            expect(resolveRequest('/demo', '../src/file.js')).to.be.resolvedTo('/src/file.js')
```
[the remark this answers is not in the document]
Comment (AviVahl) [reply packages/resolve/test/request-resolver.spec.ts]: removed

File: packages/resolve/test/request-resolver.spec.ts [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+                },
+                'style.css': EMPTY,
+                'image.jpg': EMPTY
+            })
+            const resolveRequest = createRequestResolver({ fs })
+
+            expect(resolveRequest('/', './src/file.js')).to.be.resolvedTo('/src/file.js')
+            expect(resolveRequest('/src', './data.json')).to.be.resolvedTo('/src/data.json')
+            expect(resolveRequest('/src', './typed.ts')).to.be.resolvedTo('/src/typed.ts')
+            expect(resolveRequest('/src', '../style.css')).to.be.resolvedTo('/style.css')
```
[the remark this answers is not in the document]
Comment (AviVahl) [reply packages/resolve/test/request-resolver.spec.ts]: removed

File: packages/resolve/test/request-resolver.spec.ts [hunk 8]
```diff
@@ -0,0 +1,344 @@
...
+                'another.tsx': EMPTY,
+                'style.css.ts': EMPTY,
+                'now_ignored.js': EMPTY
+            })
+            const resolveRequest = createRequestResolver({ fs, extensions: ['.ts', '.tsx'] })
+
+            expect(resolveRequest('/', './file')).to.be.resolvedTo('/file.ts')
+            expect(resolveRequest('/', './another')).to.be.resolvedTo('/another.tsx')
+            expect(resolveRequest('/src', '../style.css')).to.be.resolvedTo('/style.css.ts')
+            expect(resolveRequest('/origin', './now_ignored')).to.be.resolvedTo(null)
```
[the remark this answers is not in the document]
Comment (AviVahl) [reply packages/resolve/test/request-resolver.spec.ts]: leftovers. nice catch! fixed.

**Pull request merged**